vertical-pod-autoscaler/updater/eviction/pods_eviction_restriction.go

```diff
@@ -0,0 +1,233 @@
+// Does not check if pod was actually evicted after eviction grace period.
+func (e *podsEvictionRestrictionImpl) Evict(podToEvict *apiv1.Pod) error {
+	cr, present := e.podsCreators[getPodId(podToEvict)]
+	if !present {
+		return fmt.Errorf("pod not suitable for eviction %v : not in replicated pods map", podToEvict.Name)
+	}
+	if e.evictionBudget[cr] < 1 {
+		return fmt.Errorf("cannot evict pod %v : eviction budget exceeded", podToEvict.Name)
+	}
+
+	eviction := &policyv1.Eviction{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: podToEvict.Namespace,
+			Name:      podToEvict.Name,
+		},
+	}
+	err := e.client.CoreV1().Pods(podToEvict.Namespace).Evict(eviction)
+	if err != nil {
+		glog.Errorf("failed to evict pod %s, error: %v", podToEvict.Name, err)
+		return err
+	}
+	e.evictionBudget[cr] = e.evictionBudget[cr] - 1
+	return nil
+}
+
+// NewPodsEvictionRestrictionFactory creates PodsEvictionRestrictionFactory
+func NewPodsEvictionRestrictionFactory(client kube_client.Interface, minReplicas int, evictionToleranceFraction float64) PodsEvictionRestrictionFactory {
+	return &podsEvictionRestrictionFactoryImpl{client: client, minReplicas: minReplicas, evictionToleranceFraction: evictionToleranceFraction}
+}
+
+// NewPodsEvictionRestriction creates PodsEvictionRestriction for given set of pods.
+func (f *podsEvictionRestrictionFactoryImpl) NewPodsEvictionRestriction(pods []*apiv1.Pod) PodsEvictionRestriction {
+	// We can evict pod only if it is a part of replica set
+	// For each replica set we can evict only a fraction of pods.
+	// Evictions may be later limited by pod disruption budget if configured.
+
+	livePods := make(map[podReplicaCreator][]*apiv1.Pod)
+
+	for _, pod := range pods {
+		creator, err := getPodReplicaCreator(pod)
+		if err != nil {
+			glog.Errorf("failed to obtain replication info for pod: %v", pod.Name, err)
+			continue
+		}
+		if creator == nil {
+			glog.V(1).Infof("pod not replicated %v", pod.Name)
+			continue
+		}
+		livePods[*creator] = append(livePods[*creator], pod)
+	}
+
+	podsCreators := make(map[string]podReplicaCreator)
+	creatorsEvictionBudget := make(map[podReplicaCreator]int)
+	for creator, replicas := range livePods {
+		actual := len(replicas)
+		if actual < f.minReplicas {
+			glog.V(1).Infof("too few replicas for %v %v/%v. Found %v live pods",
+				creator.Kind, creator.Namespace, creator.Name, actual)
+			continue
+		}
+		var configured int
+		if creator.Kind == "Job" {
+			// Job has no replicas configuration, so we will use actual number of live pods as replicas count.
+			configured = actual
+		} else {
+			var err error
+			configured, err = getReplicaCount(creator, f.client)
+			if err != nil {
+				glog.Errorf("failed to obtain replication info for %v %v/%v. %v",
+					creator.Kind, creator.Namespace, creator.Name, err)
+				continue
+			}
+		}
+
+		evictionTolerance := int(float64(configured) * f.evictionToleranceFraction)
```

Comment:
If i round it up to nearest int we may evict too many pods.
Tolerance is the number of pods we safely can evict, so that evictionToleranceFraction is not exceeded.  Casting float to int ensures we will not evict too many. If in some cases floating point arithmetic will result in not equal int number (i.e. 1.999999.. instead of 2.0) I think it's still better to truncate number to 1 and evict less pods. We will evict rest in next updater execution round.